sbi/utils/sbiutils.py

```diff
@@ -265,6 +284,171 @@ def sample_posterior_within_prior(
+            with the maximum ratio of the `potential_fn / proposal` ratio. A higher
+            value will ensure that the samples are indeed from the correct
+            distribution, but will increase the fraction of rejected samples and thus
+            computation time.
+
+    Returns:
+        Accepted samples and acceptance rate as scalar Tensor.
+    """
+
+    find_max = proposal.sample((num_samples_to_find_max,))
+
+    # Define a potential as the ratio between target distribution and proposal.
+    def potential_over_proposal(theta):
+        return potential_fn(theta) - proposal.log_prob(theta)
+
+    # Search for the maximum of the ratio.
+    _, max_log_ratio = optimize_potential_fn(
+        potential_fn=potential_over_proposal,
+        inits=find_max,
+        dist_specifying_bounds=proposal,
+        num_iter=100,
+        learning_rate=0.01,
+        num_to_optimize=max(1, int(num_samples_to_find_max / 10)),
+        show_progress_bars=False,
+    )
+
+    if m < 1.0:
+        warnings.warn("A value of m < 1.0 will lead to systematically wrong results.")
+
+    class ScaledProposal:
+        def __init__(self, proposal: Any, max_log_ratio: float, log_m: float):
+            self.proposal = proposal
+            self.max_log_ratio = max_log_ratio
+            self.log_m = log_m
+
+        def sample(self, sample_shape: torch.Size, **kwargs) -> Tensor:
+            """
+            Samples from the `ScaledProposal` are samples from the `proposal`.
+            """
+            return self.proposal.sample((sample_shape,), **kwargs)
+
+        def log_prob(self, theta: Tensor, **kwargs) -> Tensor:
+            """
+            The log-prob is scaled such that the proposal is always above the potential.
+            """
+            return (
+                self.proposal.log_prob(theta, **kwargs)
+                + self.max_log_ratio
+                + self.log_m
+            )
+
+    proposal = ScaledProposal(proposal, max_log_ratio, torch.log(torch.as_tensor(m)))
+
+    with torch.no_grad():
+        # Progress bar can be skipped, e.g. when sampling after each round just for
+        # logging.
+        pbar = tqdm(
+            disable=not show_progress_bars,
+            total=num_samples,
+            desc=f"Drawing {num_samples} posterior samples",
+        )
+
+        num_sampled_total, num_remaining = 0, num_samples
+        accepted, acceptance_rate = [], float("Nan")
+        leakage_warning_raised = False
+
+        # To cover cases with few samples without leakage:
+        sampling_batch_size = min(num_samples, max_sampling_batch_size)
+        while num_remaining > 0:
+
+            # Sample and reject.
+            candidates = proposal.sample(sampling_batch_size).reshape(
+                sampling_batch_size, -1
+            )
+
+            if binary_rejection_criterion:
+                # SNPE-style rejection-sampling when the proposal is the neural net.
+                candidates = candidates.cpu()
```

Comment:
the `.cpu()` usage here is a bit confusing. In general we only need to move to cpu when comparing to the prior because the prior can't live / be moved on the device. so maybe we can reduce the usage of `cpu()` somehow by using this assumption?

Reply:
Good point, fixed it!